### ISSUE TITLE
Some questions

### ISSUE THREAD

**Issue by zhanglonghao1992** [batmanlab/MSPC#1]
When training with multi GPUS (setting `gpu_ids=0,1,2,3,4,5,6,7`), the following error will be reported:

> RuntimeError: Expected tensor for argument #1 'input' to have the same device as tensor for argument #2 'weight'; but device 1 does not equal 0 (while checking arguments for cudnn_convolution)

**Comment by zhanglonghao1992** [comment]
@xuyanwu Besides multi-GPU error, there are too many bugs in this code. Did you check it carefully before release?

**Comment by xuyanwu** [comment]
Hi,

the multi-GPU error might be caused by the two adversarial training processes in our paper. I have tried to build it on the multiple-GPUs settings. I have run the code before uploading it. I will take a further look. Thanks for the reporting.

I am currently occupied with another project, I will re-upload the code very soon, stay tuned.

**Comment by xuyanwu** [comment]
Also, all of the experiments are run on a single GPU setting, I haven't tried the code on more than one GPUs.

**Comment by zhanglonghao1992** [comment]
@xuyanwu I redefined netLOC and netTPS, now it can run with multi-GPUs. I have trained on the selfie2anime dataset and found that T didn‘t work after few iters.
pert_A and pert_B become this:
![pert_B](https://user-images.githubusercontent.com/17874285/164357474-cb1fdb1a-1a21-4c69-9d31-9a7a37e35936.png)

**Comment by xuyanwu** [comment]
I saw the issue, which is caused by the constraint. Let me check the code, will get you back very soon.

**Comment by zhanglonghao1992** [comment]
@xuyanwu It's been a while. Is there any progress?  Based on my recent attempts, I doubt that T doesn't work at all, at least on some I2I tasks.

**Comment by xuyanwu** [comment]
Hi,

I just clear the issue of the code, it was the mistake of the reverse gradient and I uploaded the wrong version of maxgcpert3, now it should be good. Also, I changed the run_gcpert.sh to the selfie2anime setting. Sorry to reply too late.

Thanks for the attention to our repo. The codes are not fully cleaned, but everything would be cleaned before the cvpr2022.

Best,
Yanwu

**Comment by xuyanwu** [comment]
I won't close the issue now, let me know if it works now.

**Comment by kayhan-batmanghelich** [comment]
@zhanglonghao1992 let us know if this update resolves your issue. Thanks

**Comment by zhanglonghao1992** [comment]
@kayhan-batmanghelich @xuyanwu  
Thank you for your kindly reply. I'll try it tomorrow.

**Comment by zhanglonghao1992** [comment]
@xuyanwu Now it seems right.
By the way, to run with multi-GPUs, you can define `netLOC`, `netLOC_T` and `netTPS` as you define `G` and `D`:
```
self.netLOC = networks.define_LOC(self.target_control_point, self.opt.span_range, self.opt.bounded, self.opt.grid_size, self.gpu_ids)
self.netLOC_T = networks.define_LOC(self.target_control_point, self.opt.span_range, self.opt.bounded, self.opt.grid_size, self.gpu_ids)
```
The function `define_LOC` can be like:
```
def define_LOC(target_control_point, span_range, bounded, grid_size, gpu_ids=[]):
    GridLocNet = {
                  'unbounded': UnBoundedGridLocNet,
                  'bounded': BoundedGridLocNet,
                 }[bounded]
    netLOC = GridLocNet(grid_size, grid_size, target_control_point)

    return init_net_loc(netLOC, gpu_ids)
```
```
def init_net_loc(net, gpu_ids=[]):
    if len(gpu_ids) > 0:
        assert(torch.cuda.is_available())
        net.to(gpu_ids[0])
        net = torch.nn.DataParallel(net, gpu_ids)

    return net
```
The definition of `netTPS` is the same. And there is no need to run https://github.com/batmanlab/MSPC/blob/3370c7041fc29dfd81479828ec2da69929691c0f/train.py#L42

**Comment by xuyanwu** [comment]
Thanks for the update, will add this function. Also, If you are training on the unpaired dataset such as selfie2anime, please choose a large coefficient for identity in the training script. I have uploaded the two steps training code as described in the paper just now, which is faster. The previous one is the three steps of training. 

Feel free to open an issue again, I am closing this one now.

**Comment by xuyanwu** [comment]
@zhanglonghao1992 Thanks for the update, will add this function. Also, If you are training on the unpaired dataset such as selfie2anime, please choose a large coefficient for identity in the training script. I have uploaded the two steps training code as described in the paper just now, which is faster. The previous one is the three steps of training.

Feel free to open an issue again, I am closing this one now.

**Comment by zhanglonghao1992** [comment]
@xuyanwu Unfortunately, after several experiments with different super parameters, the performance for the face2anime task is obviously worse than pix2pix or cycle-GAN. I still can't get the results shown in the paper.

**Comment by xuyanwu** [comment]
@zhanglonghao1992 Is it for the Business Usage or the Research? If it is for research, I will provide the pretrained ckpt. Also What is your evaluated metric?

**Comment by xuyanwu** [comment]
@zhanglonghao1992 I haven't wroten the policy of the usage thus it would be dangerous to release the ckpt. IF you are using it for your research goal, I can email your the pretrained ckpt. 

Best!

**Comment by zhanglonghao1992** [comment]
@xuyanwu For the research, I just want to find out if T can really improve the effect of G. I didn't use any quantitative evaluation metrics, just through observation.
You can just release your hyper-parameters for the face2anime dataset.

**Comment by xuyanwu** [comment]
@zhanglonghao1992 That is good! The parameters have been released in the run_gcpert.sh for the specific selfie2anime dataset, which I uploaded yesterday and just ran again.

**Comment by zhanglonghao1992** [comment]
@xuyanwu I have trained with these newly released parameters and couldn't get the good result. I only changed the resolution from 128 to 256 and used multi-GPUs with batch size 32.

**Comment by xuyanwu** [comment]
@zhanglonghao1992 That could be the reason, increasing batchsize won't help the training under the cyclegan training framework, which we all modified our codes upon. 1. Cyclegan framework doesn't involve batchnorm .2 Big batchsize leads to much less training time. So, when we train the model for comparison, we won't risk increasing batchsize as well as multi-GPU. Normally, we train it with batchsize=1. However, in our paper, we used bs=4, because we have limited time for catching up on the deadline. If you pursue the best result, I would recommend trying batchsize=1.
